spatialdata/_core/_rasterize.py

```diff
@@ -0,0 +1,383 @@
+                target_height is not None,
+                target_depth is not None,
+            ]
+        )
+        == 1
+    )
+    assert set(spatial_axes) == {"x", "y"} or set(spatial_axes) == {"x", "y", "z"}
+    if "z" not in spatial_axes:
+        assert target_depth is None
+
+    x_index = spatial_axes.index("x")
+    y_index = spatial_axes.index("y")
+    w_bb = max_coordinate[x_index] - min_coordinate[x_index]
+    h_bb = max_coordinate[y_index] - min_coordinate[y_index]
+    assert w_bb > 0
+    assert h_bb > 0
+    w_to_h_bb = w_bb / h_bb
+
+    d_bb = None
+    d_to_h_bb = None
+    if "z" in spatial_axes:
+        z_index = spatial_axes.index("z")
+        d_bb = max_coordinate[z_index] - min_coordinate[z_index]
+        assert d_bb > 0
+        d_to_h_bb = d_bb / h_bb
+
+    if target_unit_to_pixels is not None:
+        target_width = w_bb * target_unit_to_pixels
+        target_height = h_bb * target_unit_to_pixels
+        if "z" in spatial_axes:
+            assert d_bb is not None
+            target_depth = d_bb * target_unit_to_pixels
+    elif target_width is not None:
+        target_height = target_width / w_to_h_bb
+        if "z" in spatial_axes:
+            assert d_to_h_bb is not None
+            target_depth = target_height * d_to_h_bb
+    elif target_height is not None:
+        target_width = target_height * w_to_h_bb
+        if "z" in spatial_axes:
+            assert d_to_h_bb is not None
+            target_depth = target_height * d_to_h_bb
+    elif target_depth is not None:
+        assert d_to_h_bb is not None
+        target_height = target_depth / d_to_h_bb
+        target_width = target_height * w_to_h_bb
+    else:
+        raise RuntimeError("Should not reach here")
+    assert target_width is not None
+    assert isinstance(target_width, float)
+    assert target_height is not None
+    assert isinstance(target_height, float)
+    return target_width, target_height, target_depth
+
+
+@singledispatch
+def rasterize(
+    data: Union[SpatialData, SpatialElement],
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+    target_unit_to_pixels: Optional[float] = None,
+    target_width: Optional[float] = None,
+    target_height: Optional[float] = None,
+    target_depth: Optional[float] = None,
+) -> Union[SpatialData, SpatialImage]:
+    raise RuntimeError("Unsupported type: {type(data)}")
+
+
+@rasterize.register(SpatialData)
+def _(
+    sdata: SpatialData,
+    axes: tuple[str, ...],
+    min_coordinate: ArrayLike,
+    max_coordinate: ArrayLike,
+    target_coordinate_system: str,
+    target_unit_to_pixels: Optional[float] = None,
+    target_width: Optional[float] = None,
+    target_height: Optional[float] = None,
+    target_depth: Optional[float] = None,
+) -> SpatialData:
+    from spatialdata import SpatialData
+
+    new_images = {}
+    for element_type in ["points", "images", "labels", "shapes"]:
```

Comment:
do we want to iterate through all element types? Since `rasterize` isn't implemented for most element types, this will often throw an error.

I think it might be better to just do images and labels and them emit a warning if there are any points/shapes.